### PR TITLE
[raftstore-v2]: check apply_scheduler before using in on_refresh_region_buckets

### LINKED ISSUE #14506: [Dynamic Regions] bucket refresh panic
## Bug Report

[2023/04/02 20:23:26.650 +00:00] [FATAL] [lib.rs:497] ["called `Option::unwrap()` on a `None` value"] [backtrace="   0: tikv_util::set_panic_hook::{{closure}}\n   1: <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call\n             at /rustc/96ddd32c4bfb1d78f0cd03eb068b1710a8cebeef/library/alloc/src/boxed.rs:2032:9\n      std::panicking::rust_panic_with_hook\n             at /rustc/96ddd32c4bfb1d78f0cd03eb068b1710a8cebeef/library/std/src/panicking.rs:692:13\n   2: std::panicking::begin_panic_handler::{{closure}}\n             at /rustc/96ddd32c4bfb1d78f0cd03eb068b1710a8cebeef/library/std/src/panicking.rs:577:13\n   3: std::sys_common::backtrace::__rust_end_short_backtrace\n             at /rustc/96ddd32c4bfb1d78f0cd03eb068b1710a8cebeef/library/std/src/sys_common/backtrace.rs:137:18\n   4: rust_begin_unwind\n             at /rustc/96ddd32c4bfb1d78f0cd03eb068b1710a8cebeef/library/std/src/panicking.rs:575:5\n   5: core::panicking::panic_fmt\n             at /rustc/96ddd32c4bfb1d78f0cd03eb068b1710a8cebeef/library/core/src/panicking.rs:65:14\n   6: core::panicking::panic\n             at /rustc/96ddd32c4bfb1d78f0cd03eb068b1710a8cebeef/library/core/src/panicking.rs:114:5\n   7: raftstore_v2::operation::bucket::<impl raftstore_v2::fsm::peer::PeerFsmDelegate<EK,ER,T>>::on_refresh_region_buckets\n   8: raftstore_v2::fsm::peer::PeerFsmDelegate<EK,ER,T>::on_msgs\n   9: <raftstore_v2::batch::store::StorePoller<EK,ER,T> as batch_system::batch::PollHandler<raftstore_v2::fsm::peer::PeerFsm<EK,ER>,raftstore_v2::fsm::store::StoreFsm>>::handle_normal\n  10: batch_system::batch::Poller<N,C,Handler>::poll\n  11: std::sys_common::backtrace::__rust_begin_short_backtrace\n  12: core::ops::function::FnOnce::call_once{{vtable.shim}}\n  13: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once\n             at /rustc/96ddd32c4bfb1d78f0cd03eb068b1710a8cebeef/library/alloc/src/boxed.rs:2000:9\n      <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once\n             at /rustc/96ddd32c4bfb1d78f0cd03eb068b1710a8cebeef/library/alloc/src/boxed.rs:2000:9\n      std::sys::unix::thread::Thread::new::thread_start\n             at /rustc/96ddd32c4bfb1d78f0cd03eb068b1710a8cebeef/library/std/src/sys/unix/thread.rs:108:17\n  14: start_thread\n  15: __clone\n"] [location=/data4/qixu/l/tikv/components/raftstore-v2/src/operation/bucket.rs:262] [thread_name=rs-3-0]

### PR DESCRIPTION
**Pull request by tonyxuqqi**
### What is changed and how it works?

Issue Number: Close #14506

What's Changed:

```commit-message
check apply_scheduler before using it in on_refresh_region_buckets.
This is to solve the race condition when the peer is just created by split meanwhile a refresh bucket is called immediately. 
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

### Release note 

```release-note
None
```

### CODE REVIEW

File: components/raftstore-v2/src/operation/command/mod.rs [hunk 1]
```diff
@@ -148,6 +148,7 @@ impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
             .apply_pool
             .spawn(async move { apply_fsm.handle_all_tasks().await })
             .unwrap();
+        fail::fail_point!("deplay_set_apply_scheduler", |_| {});
         self.set_apply_scheduler(apply_scheduler);
```
Comment (tabokie) [inline components/raftstore-v2/src/operation/command/mod.rs:152]: Why not set it before spawning?

Comment (tonyxuqqi) [reply components/raftstore-v2/src/operation/command/mod.rs:152]: Is there any benefit to put it before spawning? The spawning itself is not going to take long time either.
The key here is to delay the set_apply_scheduler being called, so making this fail_point just prior to it is more precise. Though, in fact the delay could happen at any place when the OS thread is switched.

Comment (tonyxuqqi) [reply components/raftstore-v2/src/operation/command/mod.rs:152]: > Is there any benefit to put it before spawning? The spawning itself is not going to take long time either.
> The key here is to delay the set_apply_scheduler being called, so making this fail_point just prior to it is more precise. Though, in fact the delay could happen at any place when the OS thread is switched.

Is there any benefit to put it before spawning? The spawning itself is not going to take long time either.
The key here is to delay the set_apply_scheduler being called, so making this fail_point just prior to it is more precise. Though, in fact the delay could happen at any place when the OS thread is switched.

Comment (tonyxuqqi) [reply components/raftstore-v2/src/operation/command/mod.rs:152]: > Why not set it before spawning?

Is there any benefit to put it before spawning? The spawning itself is not going to take long time either.
The key here is to delay the set_apply_scheduler being called, so making this fail_point just prior to it is more precise. Though, in fact the delay could happen at any place when the OS thread is switched.

Comment (tabokie) [reply components/raftstore-v2/src/operation/command/mod.rs:152]: Forget it, I thought it's a race between `Apply::handle_all_tasks`.

**Review by tabokie — APPROVED** [review]

File: components/raftstore-v2/tests/failpoints/test_bucket.rs [hunk 2]
```diff
@@ -0,0 +1,58 @@
+// Copyright 2022 TiKV Project Authors. Licensed under Apache-2.0.
```
Comment (bufferflies) [inline components/raftstore-v2/tests/failpoints/test_bucket.rs:1]:
```suggestion
// Copyright 2023 TiKV Project Authors. Licensed under Apache-2.0.
```

**Review by bufferflies — APPROVED** [review]

**Pull request merged**